File: sort/merge_sort/go/merge_sort.go
package main

import "fmt"

func main() {
	arr := []int{5,15,154,1,3,2,4,8,9,5,8,4,5,6,3,2,4,5}

	fmt.Printf("Unsorted: %v\n", arr)
	fmt.Printf("Sorted: %v\n", Sort(arr))
}

func merge(left []int, right []int) []int {
	var arr = make([]int, len(left) + len(right))
	var leftIndex = 0
	var rightIndex = 0

	for leftIndex < len(left) && rightIndex < len(right) {

		if left[leftIndex] <= right[rightIndex] {
			arr[leftIndex+rightIndex] = left[leftIndex]
			leftIndex++
		} else {
			arr[leftIndex+rightIndex] = right[rightIndex]
			rightIndex++
		}

	}

	for ; leftIndex < len(left); leftIndex++ {
		arr[leftIndex+rightIndex] = left[leftIndex]
	}

	for ; rightIndex < len(right); rightIndex++ {
		arr[leftIndex+rightIndex] = right[rightIndex]
	}

	return arr
}

func Sort(items []int) []int {
	if len(items) < 2 {
		return items
	}

	middle := len(items) / 2

	a := Sort(items[:middle])
	b := Sort(items[middle:])

	return merge(a, b)
}

File: sort/merge_sort/JavaScript/MergeSort.js
/**
 * @param {Array} arr
 */
function mergeSort(arr)
{
    if (arr.length < 2)
        return arr;
 
    var mid = Math.floor(arr.length / 2);
    var left   = arr.slice(0, mid);
    var right  = arr.slice(mid, arr.length);
 
    return merge(mergeSort(left), mergeSort(right));
}
 
function merge(left, right)
{
    var result = [];
 
    while (left.length && right.length) {
        if (left[0] <= right[0]) {
            result.push(left.shift());
        } else {
            result.push(right.shift());
        }
    }
 
    while (left.length)
        result.push(left.shift());
 
    while (right.length)
        result.push(right.shift());
 
    return result;
}
 
module.exports = mergeSort;
